controllers/cluster_controller.go

```diff
@@ -18,34 +18,230 @@ package controllers
 
 import (
 	"context"
+	"path"
+	"sync"
+	"time"
 
 	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/cluster-api/pkg/controller/external"
+	capierrors "sigs.k8s.io/cluster-api/pkg/errors"
+	"sigs.k8s.io/cluster-api/pkg/util"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
 
-	clusterv1alpha2 "sigs.k8s.io/cluster-api/api/v1alpha2"
+const (
+	// deleteRequeueAfter is how long to wait before checking again to see if the cluster still has children during
+	// deletion.
+	deleteRequeueAfter = 5 * time.Second
 )
 
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
 	client.Client
 	Log logr.Logger
+
+	controller       controller.Controller
+	recorder         record.EventRecorder
+	externalWatchers sync.Map
 }
 
-// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get;update;patch
+func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.Cluster{}).
+		Build(r)
+
+	r.controller = c
+	r.recorder = mgr.GetEventRecorderFor("cluster-controller")
+	return err
+}
 
-func (r *ClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
+func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
 	_ = r.Log.WithValues("cluster", req.NamespacedName)
```

Comment:
you can get rid of this if you're not using it.

Reply:
We'll be using, in a later iteration I'll switch to klogr :)